### ISSUE TITLE
Generalization of Lukas Kande Method

### ISSUE THREAD

**Issue by sguttikon** [placeforyiming/CVPR21-Deep-Lucas-Kanade-Homography#4]
The key advantage of this method, as mentioned in paper, is the learned feature maps - DLKFM is invariant to appearance change while keeping brightness consistency (which is essential for Lukas Kanade) thus the color difference is very small between template and input image.

However, when I tried for same Great Boston area (as used in paper) but for 2017 vs 2019 (instead of 2018 vs 2019) even though feature maps for template and input visually seems to satisfy brightness consistency assumption and initial guess for iteration was MSE~15 (which should be makes it easy for convergence), the Lukas Kanade method performed poorly to align the image.

Above is very common scenario for real world drone vs satellite image since the weather conditions change.

![Slide22](https://user-images.githubusercontent.com/4303534/126123686-4672433a-53f0-40ed-bbb1-b5ccf0d2b757.jpg)

**Comment by placeforyiming** [comment]
> The key advantage of this method, as mentioned in paper, is the learned feature maps - DLKFM is invariant to appearance change while keeping brightness consistency (which is essential for Lukas Kanade) thus the color difference is very small between template and input image.
> 
> However, when I tried for same Great Boston area (as used in paper) but for 2017 vs 2019 (instead of 2018 vs 2019) even though feature maps for template and input visually seems to satisfy brightness consistency assumption and initial guess for iteration was MSE~15 (which should be makes it easy for convergence), the Lukas Kanade method performed poorly to align the image.
> 
> Above is very common scenario for real world drone vs satellite image since the weather conditions change.
> 
> ![Slide22](https://user-images.githubusercontent.com/4303534/126123686-4672433a-53f0-40ed-bbb1-b5ccf0d2b757.jpg)

Hi Suresh,

As we discussed before, the image pair of 2018 vs 2019 looks more similar to our training data. This may be why the pre-trained checkpoint works well on this pair. On the contrary, the pair 2017 vs 2019 looks different with the training date of the provided checkpoint.

One possible reason we discussed before is the brightness constancy. In our paper, we have a Siamese network with f(x) for input and f'(x') for the template image. If the input image looks different, the feature map will become f(x+delta), f'(x'). The f(x) and f'(x') have exactly the same brightness value on extract feature patterns, but f(x+delta) will be different if the delta is large. Though the feature pattern looks the same visually.

My suggestion is to use the new image pair, like 2017 vs 2019 to finetune the checkpoint a little bit, to let the network adapt to the delta. The LK is a very subtle algorithm, the benefit is the potential sub-pixel accuracy, the drawback is the high requirement of the image pair.

You can try first, if you need further help, I may have time to try the code in September. I have a couple of deadlines in the coming 2 months, so may have time to work on something else around mid of September.

Best
Yiming

**Comment by sguttikon** [comment]
Hi Yiming, 

Yes, I tried to over-fit the network for single drone and corresponding satellite image (train/eval on same image) from scratch for 1k epochs 'train_level_three.py'. As you can see, the feature map is not brightness consistent for real images i.e road, grass features look different in images. And, you can imagine the LK method won't align image in this case.

I also tried to fine-tune for 100 drone-sat image pairs (more image pairs required manual effort) and similar problem is observed.

Preprocessing/Date generation step:
1. Using this [method](https://www.mathworks.com/help/images/registering-an-aerial-photo-to-an-orthophoto.html) first I obtained gt homography for drone-sat 192x192 image pair.
2. Using process as mentioned in 'generate_train&test_sample.py' i obtained drone 128x128 image, and corner points.

Do you have any suggestions to make the feature map consistent ?

best regards
Suresh
 
![Slide26](https://user-images.githubusercontent.com/4303534/126285605-2dc35374-ac78-41b8-a744-4bbfee5de8a5.jpg)

**Comment by placeforyiming** [comment]
> Hi Yiming,
> 
> Yes, I tried to over-fit the network for single drone and corresponding satellite image (train/eval on same image) from scratch for 1k epochs 'train_level_three.py'. As you can see, the feature map is not brightness consistent for real images i.e road, grass features look different in images. And, you can imagine the LK method won't align image in this case.
> 
> I also tried to fine-tune for 100 drone-sat image pairs (more image pairs required manual effort) and similar problem is observed.
> 
> Preprocessing/Date generation step:
> 
> 1. Using this [method](https://www.mathworks.com/help/images/registering-an-aerial-photo-to-an-orthophoto.html) first I obtained gt homography for drone-sat 192x192 image pair.
> 2. Using process as mentioned in 'generate_train&test_sample.py' i obtained drone 128x128 image, and corner points.
> 
> Do you have any suggestions to make the feature map consistent ?
> 
> best regards
> Suresh
> 
> ![Slide26](https://user-images.githubusercontent.com/4303534/126285605-2dc35374-ac78-41b8-a744-4bbfee5de8a5.jpg)

Hi Suresh,

I think you may try train from scratch for this case. train_level_one.py should be used first. If you can observe a significant decrease of the training loss, the network should start learning something meaningful. I think you can test on few images first, it may overfit, but should can tell if the method work or not.

**Comment by sguttikon** [comment]
Hi Yiming,

It seems the layers 1 to 3 are able to extract meaningful features (as you can see compared to previously), which is an good progress.
However, the regression network(s) loss seemed to not improving, even after training single image for 5k epochs. Do you have any insight about this, and any suggestions to improve it?

Feature Map (after training):
<img src="https://user-images.githubusercontent.com/4303534/127548514-6e65ff5e-5898-4dc2-8c02-206e919289f4.jpg" width="512" height="256">

![#FF8C00](https://via.placeholder.com/15/FF8C00/000000?text=+) - *level_one*, ![#1589F0](https://via.placeholder.com/15/1589F0/000000?text=+) - *level_two*, ![#8e3109](https://via.placeholder.com/15/8e3109/000000?text=+) - *level_three*
|convex loss | ssim loss |
|----------------|--------------|
|<img src="https://user-images.githubusercontent.com/4303534/127546559-ca258a06-17fa-4696-a849-21b4890bce1c.png" width="512" height="256">|<img src="https://user-images.githubusercontent.com/4303534/127546562-c3c4191b-3b0d-4ded-88c7-7e66bf690649.png" width="512" height="256">|


```
total_loss += tf.reduce_mean(tf.math.sqrt((homography_vector-gt_vector)**2))
loss_1= loss_func(input_parameters.batch_size,homography_vector,u_list,v_list)
```
![#1acdd8](https://via.placeholder.com/15/1acdd8/000000?text=+) - *regression_first*, ![#c40991](https://via.placeholder.com/15/c40991/000000?text=+) - *regression_second*, ![#0e7076](https://via.placeholder.com/15/0e7076/000000?text=+) - *regression_third*
| total loss| loss_1 |
|-------------|----------|
|<img src="https://user-images.githubusercontent.com/4303534/127547313-90bd2747-eb99-4530-8933-0808cf2b0202.png" width="512" height="256">|<img src="https://user-images.githubusercontent.com/4303534/127547328-6be3bb41-f9ec-4f31-8c1e-277253c2dd1a.png" width="512" height="256">|

**Comment by placeforyiming** [comment]
> Hi Yiming,
> 
> It seems the layers 1 to 3 are able to extract meaningful features (as you can see compared to previously), which is an good progress.
> However, the regression network(s) loss seemed to not improving, even after training single image for 5k epochs. Do you have any insight about this, and any suggestions to improve it?
> 
> Feature Map (after training):
> <img alt="" width="512" height="256" src="https://user-images.githubusercontent.com/4303534/127548514-6e65ff5e-5898-4dc2-8c02-206e919289f4.jpg">
> 
> ![#FF8C00](https://camo.githubusercontent.com/59120190fd3bf6802c4d9f3c96a05c351dbf3878d9c9bee0fa369a918a9ab509/68747470733a2f2f7669612e706c616365686f6c6465722e636f6d2f31352f4646384330302f3030303030303f746578743d2b) - _level_one_, ![#1589F0](https://camo.githubusercontent.com/2e509e3b9a17afb9b8a2f7d9a05a6f0dcdb01b7a528a8f90a9e058a0cb07b869/68747470733a2f2f7669612e706c616365686f6c6465722e636f6d2f31352f3135383946302f3030303030303f746578743d2b) - _level_two_, ![#8e3109](https://camo.githubusercontent.com/be638b2ddec8045d077c50f252cf722644764b8de28c6e70b16bc86a25814508/68747470733a2f2f7669612e706c616365686f6c6465722e636f6d2f31352f3865333130392f3030303030303f746578743d2b) - _level_three_
> 
> convex loss	ssim loss
> <img alt="" width="512" height="256" src="https://user-images.githubusercontent.com/4303534/127546559-ca258a06-17fa-4696-a849-21b4890bce1c.png">	<img alt="" width="512" height="256" src="https://user-images.githubusercontent.com/4303534/127546562-c3c4191b-3b0d-4ded-88c7-7e66bf690649.png">
> ```
> total_loss += tf.reduce_mean(tf.math.sqrt((homography_vector-gt_vector)**2))
> loss_1= loss_func(input_parameters.batch_size,homography_vector,u_list,v_list)
> ```
> 
> ![#1acdd8](https://camo.githubusercontent.com/dbc634661e4a19e5c6d1ee2fe061f9ae754a21c349657849b7ec4374d23d91b8/68747470733a2f2f7669612e706c616365686f6c6465722e636f6d2f31352f3161636464382f3030303030303f746578743d2b) - _regression_first_, ![#c40991](https://camo.githubusercontent.com/7101a8241e67b0f2aaa35881dfbba870e257ba97494db44e67e58fce87fda5df/68747470733a2f2f7669612e706c616365686f6c6465722e636f6d2f31352f6334303939312f3030303030303f746578743d2b) - _regression_second_, ![#0e7076](https://camo.githubusercontent.com/c9dd733c1cd6c307db31c5411b46120c5ef96bc95dac4383264c86f229e899fd/68747470733a2f2f7669612e706c616365686f6c6465722e636f6d2f31352f3065373037362f3030303030303f746578743d2b) - _regression_third_
> 
> total loss	loss_1
> <img alt="" width="512" height="256" src="https://user-images.githubusercontent.com/4303534/127547313-90bd2747-eb99-4530-8933-0808cf2b0202.png">	<img alt="" width="512" height="256" src="https://user-images.githubusercontent.com/4303534/127547328-6be3bb41-f9ec-4f31-8c1e-277253c2dd1a.png">

Hi Suresh,

It seems our method is working. The regression part is my implementation of other people's models. They are working on datasets used in the paper. I mean, worse than they claimed in their paper, but working. You can try to adjust part of the regression models. Or, if your GPS and IMU are good enough, you may don't need initial

**Comment by placeforyiming** [comment]
Please feel free to open it again if there are something else
